widget/embedded/src/lib/tabManager/tabManager.ts

```diff
@@ -0,0 +1,235 @@
+import type {
+  Events,
+  ID,
+  Payload,
+  TabManagerInterface,
+} from './tabManager.types';
+
+import { CHANNEL_NAME, TIMEOUT } from './tabManager.constants';
+
+export class TabManager implements TabManagerInterface {
+  private id: ID;
+  private claimed: boolean | null = null;
+  private lastCheck: number | undefined;
+  private events: Events = {};
+  private channel: BroadcastChannel;
+  private tabIDs: ID[] = [];
+
+  constructor(events: Events) {
+    this.channel = new BroadcastChannel(CHANNEL_NAME);
+    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+    this.id = Math.trunc(Math.random() * 100000);
+    this.events = events;
+    this.initEvents();
+    this.tryClaim();
+  }
+
+  forceClaim = (): void => {
+    if (!this.claimed) {
+      const message: Payload = { name: 'force-claim', id: this.id };
+      this.channel.postMessage(message);
+    }
+  };
+
+  isClaimed() {
+    return !!this.claimed;
+  }
+
+  /**
+   *
+   * Attach some event listeners to the window.
+   * It includes listening to broadcast messages, entering and exiting the page event.
+   *
+   */
+  private initEvents() {
+    const message: Payload = { name: 'open', id: this.id };
+    this.channel.postMessage(message);
+
+    this.channel.addEventListener('message', (event: MessageEvent<Payload>) => {
+      const name = event.data.name;
+
+      switch (name) {
+        case 'ping':
+          this.claimedByCurrentTab(event.data.pingAt);
+          break;
+
+        case 'pong':
+          this.alreadyClaimedByAnotherTab();
+          break;
+
+        case 'release':
+          this.updateAndClaim(event.data.id, event.data.openedTabs);
+          break;
+
+        case 'open':
+          this.saveTabID(event.data.id);
+          break;
+
+        case 'close':
+          this.removeTabId(event.data.id);
+          break;
+
+        case 'force-claim':
+          this.forceRelease(event.data.id);
+          break;
+
+        case 'force-release':
+          this.updateAndClaim(event.data.id, event.data.openedTabs);
```

Comment:
you could define a variable for event.data.id before switch to avoid duplicate code.

Reply:
As the data.id may not exist for all types, creating a variable or destructuring it is not feasible.